### PR TITLE
Consumer: Handle disconnection errors on committing offsets

### PR DESCRIPTION
**Pull request by oliakremmyda**
TL;DR: Added Logger and handle disconnection errors on committing offsets.

----

When the consumer client gets disconnected from its rafka before committing its offsets
`@redis.rpush("acks")` fails with the error, as other/new rafka has lost the details for
this client that is kept in its memory.
```
CONS No consumer registered for Client 1233XYZ
```

Upon such errors `@redis.blpop`, reconnects with the rafka and reregister the client.
Since disconnections, can be common and during consuming a message with `blpop` they are
autoresolved, we want similar behavior of self healing also during committing offsets.
However after a reconnection, it is not safe to commit anything as another consumer
might have already taken over the partition, we just log the disconnection event and leave the
offsets uncommitted for the next consume event to commit after reregister properly with blpop.

This will result to reprocessing some of the messages, but reprocessing the same message is part
of the kafka message processing specification, that guaranties at-least-one delivery of a message.
Consumers are already designed to handle this in their implementation.

If any other error is send from rafka server, it will be reraised as before to the client.

### CODE REVIEW

**Comment by oliakremmyda** [comment]
## Before the change
How to get the `CONS No consumer registered for Client` error:

Scenario:
- Create a ruby consumer that consumes a message.
- Before committing the offset of the read message sleeps for 10 secs 
- During these seconds, I close and reopen rafka to simulate the disconnection, and connecting to a different/new rafka that does not have this client in memory.  
- When the consumer tries to finally commit its offset gets the error: `CONS No consumer registered for Client` and fails.

https://github.com/user-attachments/assets/915d0071-94e3-40ad-b660-602f810a95c7

## After these changes
Scenario # 1:
- Create a ruby consumer that consumes a message.
- Before committing the offset of the read message sleeps for 10 secs 
- During these seconds, I close and reopen rafka to simulate the disconnection, and connecting to a different/new rafka that does not have this client in memory.  
- When the consumer tries to finally commit its offset prints the warning `CONS Cannot commit offsets, connection with rafka was temporarly lost.` and returns without error and without committing the read message. It will reread the same message afterwards

https://github.com/user-attachments/assets/c8457454-32bc-4a4d-a8d4-2abf39a57416

Scenario # 2
- Create a ruby consumer that consumes a message.
- Before committing the offset of the read message sleeps for 10 secs 
- During these seconds, I close and reopen rafka to simulate the disconnection, and connecting to a different/new rafka that does not have this client in memory.  
- When the consumer tries to finally commit its offset prints the warning `CONS Cannot commit offsets, connection with rafka was temporarly lost.` and returns without error and without committing the read message.
- Another consumer is initated and takes over the partition while the first waits to take control.
- The new consumer reads the rest of the messages and the first one returns empty from read.

https://github.com/user-attachments/assets/e7a4a79d-b647-400f-83ac-76bd05eed321

File: lib/rafka/consumer.rb [hunk 1]
```diff
@@ -185,7 +192,15 @@ def commit(*msgs)
...
         po.each do |partition, offset|
           Rafka.wrap_errors do
-            @redis.rpush("acks", "#{topic}:#{partition}:#{offset}")
+            begin
+              @redis.rpush("acks", "#{topic}:#{partition}:#{offset}")
+            rescue Redis::CommandError => e
+              if e.message.start_with?("CONS No consumer registered for Client")
+                @logger.warn("CONS Cannot commit offsets, connection with rafka was temporarly lost.")
+              else
+                raise e
```
Comment (avgerin0s) [inline lib/rafka/consumer.rb:201]: Since this is a library I wouldn't go as far as setting a logger here.

A more un-opinionated approach, imho, is to create the following error in `lib/error.rb` and re-raise it letting the library user handle it and possibly log whatever they want:

```
class UnregisteredConsumerError < ConsumeError; end
```

```
rescue Redis::CommandError => e
  if e.message.start_with?("CONS No consumer registered for Client")
     raise UnregisteredConsumerError.new(e.message)
  else
    raise e
  end
```

what do you think?

Comment (MrGoumX) [reply lib/rafka/consumer.rb:201]: We could also utilize the `wrap_errors` function in order not to redefine `Redis::CommandError` error handlings again

```ruby
po.each do |partition, offset|
  Rafka.wrap_errors do
    @redis.rpush("acks", "#{topic}:#{partition}:#{offset}")
  end
rescue ConsumeError => e
  raise e unless e.message.start_with?("No consumer registered for Client")

  raise UnregisteredConsumerError.new(e.message)
end
```

Comment (oliakremmyda) [reply lib/rafka/consumer.rb:201]: My intention was to address this at one place for all the requests.  If we don't want that, I don't see any value of reraising the error with a different type here. The consumer will have to catch it and handle it in every consume call. In this is the case, I would leave the library as is and filter the original error in client's side every time it calls consume, or commit interfaces.

**Comment by oliakremmyda** [comment]
We've discussed to handle this case client side, closing this.
